packages/cli/src/lib/features.ts

```diff
@@ -0,0 +1,163 @@
+export function getFeaturesMetadata(
+  project: Project,
+  role: PackageRole,
+  dir: string,
+  entryPoints: EntryPoint[],
+): BackstagePackageFeature[] {
+  if (!targetPackageRoles.includes(role)) {
+    return [];
+  }
+
+  return entryPoints
+    .flatMap(entryPoint => getEntryPointExports(entryPoint, project, dir))
+    .filter(isDefaultExport)
+    .filter(isTargetFeatureType)
+    .map(toBackstagePackageFeature);
+}
+
+type EntryPointExport = {
+  name: string;
+  location: string;
+  type: BackstagePackageFeatureType;
+};
+
+// Returns all exports (default and named) from an entry point
+// that are valid Backstage package features
+export function getEntryPointExports(
+  { mount: location, path, ext }: EntryPoint,
+  project: Project,
+  dir: string,
+): EntryPointExport[] {
+  const fullFilePath = resolvePath(dir, path);
+  const sourceFile = project.getSourceFile(fullFilePath);
+
+  if (!sourceFile || (ext !== '.ts' && ext !== '.tsx')) {
+    return [];
+  }
+
+  const exports = [];
+
+  for (const exportSymbol of sourceFile.getExportSymbols()) {
+    const declaration = exportSymbol.getDeclarations()[0];
+    const exportName = declaration.getSymbol()?.getName();
+    let exportType: Type<ts.Type> | undefined;
+
+    if (declaration) {
+      if (declaration.isKind(SyntaxKind.ExportAssignment)) {
+        exportType = declaration.getExpression().getType();
+      } else if (declaration.isKind(SyntaxKind.ExportSpecifier)) {
+        if (!declaration.isTypeOnly()) {
+          exportType = declaration.getType();
+        }
+      } else if (declaration.isKind(SyntaxKind.VariableDeclaration)) {
+        exportType = declaration.getType();
+      }
+    }
+
+    if (exportName && exportType) {
+      const $$type = getBackstagePackageFeature$$TypeFromType(exportType);
+
+      if ($$type) {
+        exports.push({
+          name: exportName,
+          location,
+          type: $$type,
+        });
+      }
+    }
+  }
+
+  return exports;
+}
+
+// Given a TS type, returns the Backstage package feature $$type value
+function getBackstagePackageFeature$$TypeFromType(
+  type: Type,
+): BackstagePackageFeatureType | null {
+  // Returns the concrete type of a generic type
+  const exportType = type.getTargetType() ?? type;
+
+  for (const property of exportType.getProperties()) {
+    if (property.getName() === '$$type') {
+      const $$type = property
+        .getValueDeclaration()
+        ?.getText()
+        .match(/(?<type>@backstage\/\w+)/)?.groups?.type;
+
+      if ($$type && isValidPackageFeatureType($$type)) {
+        return $$type;
+      }
+    }
+  }
+
+  return null;
+}
+
+// Returns whether an export is the default export
+function isDefaultExport({ name }: EntryPointExport): boolean {
+  return name === 'default';
+}
+
+// Returns whether an export is a valid Backstage package feature type
+function isTargetFeatureType({ type }: EntryPointExport): boolean {
+  return targetFeatureTypes.includes(type);
+}
```

Comment:
Thinking these two could be inlined tbh

Reply:
They could, I find it's helpful when reading their use to see what's going on. 

```ts
return entryPoints
    .flatMap(entryPoint => getEntryPointExports(entryPoint, project, dir))
    .filter(isDefaultExport)
    .filter(isTargetFeatureType)
    .map(toBackstagePackageFeature);
```